network/ip/bridge.go

```diff
@@ -0,0 +1,43 @@
+// Copyright 2023 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ip
+
+import "fmt"
+
+// An IPBridge forwards IP traffic bidirectionally between two IPDevices.
+type IPBridge struct {
```

Comment:
@fortuna although this PR is still in a very early stage, I think it would be helpful if we should agree on the interface designs before the actual implementation. So please feel free to comment on this PR while I'm still working on it.

Reply:
`io.Copy` function will do the work until either one is closed, it does not provide re-plug features, for example, we can unplug tun2socks from an old tun device and replug that into a new one. This might not be a requirement for Outline now, I'm just trying to make it flexible.

Also using a type has some semantic meanings, the caller will know that it will bridge the two endpoints. Although two `io.Copy` calls is pretty simple, it does not express the intention clearly behind the scene.

Reply:
The device is independent of the bridge. If someone wants to replug, it should be easy for them to implement that.

In practice, I don't think a Bridge type will help much, and may make people think they need it. I'd like to see that being used in practice with concrete examples to have a better idea.

We should convert our outline-go-tun2socks code to use the new interfaces when they are ready to exercise them and iterate.

Reply:
A quick sample would be like:

```
br := NewIPBridge(tun, t2s)
defer br.Destroy()
go br.Open()
```

versus

```
go io.Copy(tun, t2s)
io.Copy(t2s, tun)
```

Also it's not easy to implement "unplug" if we are using `io.Copy`, because `io.Copy` will do the work until EOF. But if we implement IPDevice and the IPBridge carefully, we can use `Context`'s `cancel()` to terminate the process.